Investigate why Clang/LLVM UBSan doesn't link internally

While merging #3452 (internal MSVC-PR-457627), I had to disable the Clang/LLVM UBSan configurations for the internal test harness, as they fail to link (with linker errors about support machinery). We use Clang/LLVM via an internally captured "NuGet package" with a subset of binaries (but most of the support libs, AFAIK) to avoid excessive space consumption. At the time, I am unsure why UBSan  doesn't work in the internal environment. It could be that we need to add more libs to the NuGet package, or it could be due to PATH ordering issues where we're mixing up MSVC and Clang/LLVM's support libs.

Thanks to @zacklj89 for reminding me to file this followup issue instead of being a bad kitty :smirk_cat: and silently working around it forever.

I'm now seeing this with VS 2022 17.7 Preview 1 and Clang 16:

```
code: FAIL
name: std :: tests/GH_002334_branchless_clamp:27
output: Build setup steps:
Build steps:
Build step failed unexpectedly.
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-cl.EXE" "D:\GitHub\STL\tests\std\tests\GH_002334_branchless_clamp\test.cpp" "-ID:\GitHub\STL\out\x64\out\inc" "-ID:\GitHub\STL\llvm-project\libcxx\test\support" "-ID:\GitHub\STL\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/w14640" "/Zc:threadSafeInit-" "-fno-ms-compatibility" "-fno-delayed-template-parsing" "-Wno-unqualified-std-cast-call" "/EHsc" "/MT" "/std:c++latest" "/permissive-" "/fp:strict" "-fsanitize=undefined" "-fno-sanitize-recover=undefined" "-m64" "-FeD:\GitHub\STL\out\x64\tests\std\tests\GH_002334_branchless_clamp\Output\27\GH_002334_branchless_clamp.exe" "-link" "-LIBPATH:D:\GitHub\STL\out\x64\out\lib\amd64" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.37.32705\lib\x64" "/MANIFEST:EMBED"
Exit Code: 1120
Standard Output:
--
   Creating library D:\GitHub\STL\out\x64\tests\std\tests\GH_002334_branchless_clamp\Output\27\GH_002334_branchless_clamp.lib and object D:\GitHub\STL\out\x64\tests\std\tests\GH_002334_branchless_clamp\Output\27\GH_002334_branchless_clamp.exp
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_printf.cpp.obj) : error LNK2019: unresolved external symbol "bool __cdecl __coe_win::ContinueOnError(void)" (?ContinueOnError@__coe_win@@YA_NXZ) referenced in function "void __cdecl __sanitizer::SharedPrintfCodeNoBuffer(bool,char *,int,char const *,char *)" (?SharedPrintfCodeNoBuffer@__sanitizer@@YAX_NPEADHPEBD1@Z)
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_common.cpp.obj) : error LNK2001: unresolved external symbol "bool __cdecl __coe_win::ContinueOnError(void)" (?ContinueOnError@__coe_win@@YA_NXZ)
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_symbolizer_win.cpp.obj) : error LNK2001: unresolved external symbol "bool __cdecl __coe_win::ContinueOnError(void)" (?ContinueOnError@__coe_win@@YA_NXZ)
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_stacktrace_libcdep.cpp.obj) : error LNK2001: unresolved external symbol "bool __cdecl __coe_win::ContinueOnError(void)" (?ContinueOnError@__coe_win@@YA_NXZ)
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_printf.cpp.obj) : error LNK2019: unresolved external symbol "void __cdecl __coe_win::RawWrite(char const *)" (?RawWrite@__coe_win@@YAXPEBD@Z) referenced in function "void __cdecl __sanitizer::SharedPrintfCodeNoBuffer(bool,char *,int,char const *,char *)" (?SharedPrintfCodeNoBuffer@__sanitizer@@YAX_NPEADHPEBD1@Z)
clang_rt.ubsan_standalone-x86_64.lib(sanitizer_stacktrace_libcdep.cpp.obj) : error LNK2019: unresolved external symbol "void __cdecl __coe_win::PrintStack(struct __sanitizer::StackTrace const *)" (?PrintStack@__coe_win@@YAXPEBUStackTrace@__sanitizer@@@Z) referenced in function "public: void __cdecl __sanitizer::StackTrace::Print(void)const " (?Print@StackTrace@__sanitizer@@QEBAXXZ)
D:\GitHub\STL\out\x64\tests\std\tests\GH_002334_branchless_clamp\Output\27\GH_002334_branchless_clamp.exe : fatal error LNK1120: 3 unresolved externals
--
Standard Error:
--
clang-cl: error: linker command failed with exit code 1120 (use -v to see invocation)
--
```

--- a/stl/inc/__msvc_string_view.hpp
+++ b/stl/inc/__msvc_string_view.hpp
@@ -360,18 +360,23 @@ public:
     _NODISCARD static _CONSTEXPR17 int compare(_In_reads_(_Count) const _Elem* const _First1,
         _In_reads_(_Count) const _Elem* const _First2, const size_t _Count) noexcept /* strengthened */ {
         // compare [_First1, _First1 + _Count) with [_First2, ...)
-#if _HAS_CXX17
-        if (_STD _Is_constant_evaluated()) {
-            if constexpr (is_same_v<_Elem, wchar_t>) {
-                return __builtin_wmemcmp(_First1, _First2, _Count);
+#if _USE_STD_VECTOR_ALGORITHMS
+        if (!_STD _Is_constant_evaluated()) {
+            // TRANSITION, GH-2289: Use vectorized algorithms for better performance than __builtin_wmemcmp.
+            const size_t _Pos = _Mismatch_vectorized<sizeof(_Elem)>(_First1, _First2, _Count);
+            if (_Pos == _Count) {
+                return 0;
             } else {
-                return _Primary_char_traits::compare(_First1, _First2, _Count);
+                return _First1[_Pos] < _First2[_Pos] ? -1 : +1;
             }
         }
-#endif // _HAS_CXX17
+#endif // ^^^ _USE_STD_VECTOR_ALGORITHMS ^^^
 
-        return _CSTD wmemcmp(
-            reinterpret_cast<const wchar_t*>(_First1), reinterpret_cast<const wchar_t*>(_First2), _Count);
+        if constexpr (is_same_v<_Elem, wchar_t>) {
+            return __builtin_wmemcmp(_First1, _First2, _Count);
+        } else {
+            return _Primary_char_traits::compare(_First1, _First2, _Count);
+        }
     }
 
     _NODISCARD static _CONSTEXPR17 size_t length(_In_z_ const _Elem* _First) noexcept /* strengthened */ {

--- a/tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
+++ b/tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
@@ -23,8 +23,6 @@
 using namespace std;
 using namespace chrono;
 
-constexpr auto intmax_max = numeric_limits<intmax_t>::max();
-
 template <typename CharT>
 [[nodiscard]] constexpr const CharT* choose_literal(const char* const str, const wchar_t* const wstr) noexcept {
     if constexpr (is_same_v<CharT, char>) {
@@ -252,8 +250,6 @@ void empty_braces_helper(
 
 template <typename CharT>
 void test_duration_formatter() {
-    using LongRatio = ratio<intmax_max - 1, intmax_max>;
-
     empty_braces_helper(seconds{5}, STR("5s"));
     empty_braces_helper(minutes{7}, STR("7min"));
     empty_braces_helper(hours{9}, STR("9h"));
@@ -266,8 +262,13 @@ void test_duration_formatter() {
     empty_braces_helper(duration<int, ratio<22, 7>>{40}, STR("40[22/7]s"));
     empty_braces_helper(duration<int, ratio<53, 101>>{40}, STR("40[53/101]s"));
     empty_braces_helper(duration<int, ratio<201, 2147483647>>{40}, STR("40[201/2147483647]s"));
-    // TRANSITION, LWG-3921: duration_cast used in formatting may raise UB
+
+#if 0 // TRANSITION, LWG-3921: Our duration formatting constructs an hh_mm_ss, which calls duration_cast,
+      // which triggers signed integer overflow when a duration has a pathological ratio, like this:
+    constexpr auto intmax_max = numeric_limits<intmax_t>::max();
+    using LongRatio           = ratio<intmax_max - 1, intmax_max>;
     empty_braces_helper(duration<int, LongRatio>{1}, STR("1[9223372036854775806/9223372036854775807]s"));
+#endif // ^^^ disabled test due to undefined behavior ^^^
 
     // formatting small types needs to work as iostreams << VSO-1521926
     empty_braces_helper(duration<long long, atto>{123}, STR("123as"));

--- a/tests/std/tests/VSO_0000000_vector_algorithms_floats/env.lst
+++ b/tests/std/tests/VSO_0000000_vector_algorithms_floats/env.lst
@@ -46,8 +46,7 @@ PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsin
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MDd /std:c++17 /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++20 /permissive- /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MTd /std:c++latest /permissive- /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
-# TRANSITION, GH-3568
-# PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++latest /permissive- /w14640 /Zc:threadSafeInit- -fsanitize=undefined -fno-sanitize-recover=undefined --start-no-unused-arguments"
+PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++latest /permissive- /w14640 /Zc:threadSafeInit- -fsanitize=undefined -fno-sanitize-recover=undefined --start-no-unused-arguments"
 RUNALL_CROSSLIST
 *	PM_CL="/fp:strict"
 *	PM_CL="/fp:precise"

--- a/tests/std/tests/VSO_0000000_vector_algorithms_mismatch_and_lex_compare/test.cpp
+++ b/tests/std/tests/VSO_0000000_vector_algorithms_mismatch_and_lex_compare/test.cpp
@@ -172,9 +172,13 @@ namespace test_mismatch_sizes_and_alignments {
 
     template <class T, size_t Size, size_t PadSize>
     char stack_array_various_alignments_impl() {
+#if defined(__clang__) && __has_feature(undefined_behavior_sanitizer)
+        // Avoid testing misaligned inputs, which UBSan would reject.
+#else // ^^^ UBSan / no UBSan vvv
         with_pad<T, Size + 1, PadSize + 1> a = {};
         with_pad<T, Size + 1, PadSize + 1> b = {};
         assert(mismatch(begin(a.v), end(a.v), begin(b.v), end(b.v)) == make_pair(end(a.v), end(b.v)));
+#endif // ^^^ no UBSan ^^^
         return 0;
     }
 

--- a/tests/std/tests/VSO_0000000_vector_algorithms_mismatch_and_lex_compare/test.cpp
+++ b/tests/std/tests/VSO_0000000_vector_algorithms_mismatch_and_lex_compare/test.cpp
@@ -172,7 +172,13 @@ namespace test_mismatch_sizes_and_alignments {
 
     template <class T, size_t Size, size_t PadSize>
     char stack_array_various_alignments_impl() {
-#if defined(__clang__) && __has_feature(undefined_behavior_sanitizer)
+#ifdef __clang__
+#if __has_feature(undefined_behavior_sanitizer)
+#define TESTING_UBSAN
+#endif
+#endif
+
+#ifdef TESTING_UBSAN
         // Avoid testing misaligned inputs, which UBSan would reject.
 #else // ^^^ UBSan / no UBSan vvv
         with_pad<T, Size + 1, PadSize + 1> a = {};

--- a/tests/std/tests/usual_matrix.lst
+++ b/tests/std/tests/usual_matrix.lst
@@ -53,5 +53,4 @@ PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsin
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MDd /std:c++17 /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++20 /permissive- /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
 PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MTd /std:c++latest /permissive- /fp:strict /w14640 /Zc:threadSafeInit- --start-no-unused-arguments"
-# TRANSITION, GH-3568
-# PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++latest /permissive- /fp:strict /w14640 /Zc:threadSafeInit- -fsanitize=undefined -fno-sanitize-recover=undefined --start-no-unused-arguments"
+PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing -Wno-unqualified-std-cast-call /EHsc /MT /std:c++latest /permissive- /fp:strict /w14640 /Zc:threadSafeInit- -fsanitize=undefined -fno-sanitize-recover=undefined --start-no-unused-arguments"

--- a/tests/utils/stl/test/features.py
+++ b/tests/utils/stl/test/features.py
@@ -52,8 +52,7 @@ def getDefaultFeatures(config, litConfig):
         DEFAULT_FEATURES.append(Feature(name='x86'))
 
     elif litConfig.target_arch.casefold() == 'x64'.casefold():
-        # TRANSITION, GH-3568
-        # DEFAULT_FEATURES.append(Feature(name='ubsan'))
+        DEFAULT_FEATURES.append(Feature(name='ubsan'))
         DEFAULT_FEATURES.append(Feature(name='edg'))
         DEFAULT_FEATURES.append(Feature(name='arch_avx2'))
         DEFAULT_FEATURES.append(Feature(name='x64'))

